doc/md/fundamentals/3-declarations/5-type-declarations.md

```diff
@@ -0,0 +1,125 @@
+
+## Defining a type
+
+The `type` keyword assigns a name to a type and makes it reusable.
+
+```motoko no-repl
+// An alias for Nat
+type Age = Nat;
+
+// An alias for Text
+type Username = Text;
+```
+
+These types can then be used in function definitions.
+
+```motoko no-repl
+func greet(name : Username, age : Age) : Text {
+    "Hello, " # name # "! You are " # Nat.toText(age) # " years old."
+}
+```
+
+## Record types
+
+In Motoko, a type can define a structured [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) with labeled fields. Each field has a specific type, and you can access them using dot notation. Records are useful for organizing related data clearly and safely.
+
+```motoko no-repl
+// A reusable record
+type Ghost = {
+    firstName : Text;
+    lastName : Text;
+    age : Nat;
+};
+
+// An instance of Ghost
+let motoko : Ghost = {
+    firstName = "Motoko";
+    lastName = "Sentinels";
+    age = 30;
+};
+```
+
+## Variant types
+
+A type can also define variants, which represent different possible states or alternatives. Variants allow a value to be one of several labeled options, making it easy to handle data that can take multiple forms.
+
+```motoko no-repl
+// Allows only one of its variants at a time.
+type Status = {
+    #Active;
+    #Inactive;
+
+// Carries an additional Text value.
+    #Banned : Text;
+};
+
+let userStatus : Status = #Active;
+let bannedUser : Status = #Banned("Violation of rules");
+```
+
+## Parameterized (generic) types
+
+Type declarations can be parameterized, allowing them to work flexibly with multiple types while ensuring type safety. This lets you create generic and reusable type definitions that adapt to different data types as needed.
+
+```motoko no-repl
+// `Box<T> is a generic type where T represents any type.
+type Box<T> = {
+    value: T;
+};
+
+// numberBox stores a Nat and textBox stores a Text.
+let numberBox : Box<Nat> = { value = 42 };
+let textBox : Box<Text> = { value = "Hello" };
+```
+
+## Recursive types and productivity
+
+Motoko allows recursive type definitions as long as they are productive. This means that any recursion in a type must pass through a constructor (such as an option (`?`), a variant, or a record field) before referring back to itself.
+
+#### Productive recursive type example
+
+```motoko no-repl
+type List<T> = ?(T, List<T>);
+```
+
+`List<T>` defines a linked list where each node holds a value of type `T` and points to either another `List<T>` or `null` to mark the end. Since the recursion passes through a constructor (`?`), this type is productive and accepted by the compiler.
+
+#### Non-productive recursive type example
+
+```motoko no-repl
+type C = C; // This definition infinitely refers to itself
+```
+
+This type never resolves to a concrete value because it lacks a constructor in its recursion; it is non-productive and rejected by the compiler.
```

Comment:
```suggestion
This type never produces a concrete type when expanded because its definition is too cyclic; it is reported as a _non-productive_  type definition and rejected by the compiler.
```